### PR TITLE
Hierarchical nSpec

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-3871

## Changes
- Support hierarchically nspec eg. "foo.bar.serviceName"

### CODE REVIEW

File: src/foam/nanos/boot/Boot.java [hunk 1]
```diff
@@ -107,6 +117,34 @@ public Object create(X x) {
...
+        }
+        pos = next;
+      }
+      var serviceName = sp.getName().substring(pos + 1);
+      NSpecFactory factory = new NSpecFactory((ProxyX) x, sp);
+      factories_.put(sp.getName(), factory);
+      logger.info("Registering", sp.getName());
+      x.putFactory(serviceName, factory);
+    }
+
```
Comment (jlhughes) [inline src/foam/nanos/boot/Boot.java:137]: So this is building a context named:
foo  and 
foo.bar  with foo as its parent.
@kgrgreer It was my understanding to have 
foo
bar

File: src/foam/nanos/boot/Boot.java [hunk 2]
```diff
@@ -64,12 +64,22 @@ public Boot(String datadir) {
...
+    // Hierarchically nspecs
+    var hnSpecs   = new ArrayList<NSpec>();
     for ( int i = 0 ; i < l.size() ; i++ ) {
       NSpec sp = (NSpec) l.get(i);
       if ( ! sp.getEnabled() ) {
         logger.info("Disabled", sp.getName());
         continue;
       }
+
+      // Skip hierarchically nspecs to be loaded later since the service would
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: I don't think we need to skip. Just create sub-contexts as needed if missing.

File: src/foam/nanos/boot/Boot.java [hunk 3]
```diff
@@ -107,6 +117,34 @@ public Object create(X x) {
       }
     });
 
+    // Load hierarchically nspecs eg. "foo.bar.serviceName":
+    // 1. Create "foo" proxy sub context from the root context
+    // 2. Create "foo.bar" proxy sub context from "foo" sub context
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: Should be a "bar" sub-context in the "foo" context.

File: src/foam/nanos/boot/Boot.java [hunk 4]
```diff
@@ -107,6 +117,34 @@ public Object create(X x) {
...
+    // 3. Create "serviceName" object in "foo.bar" sub context
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: In the x.get("foo").get("bar") sub-context.

File: src/foam/nanos/boot/Boot.java [hunk 5]
```diff
@@ -107,6 +117,34 @@ public Object create(X x) {
...
+    for ( var sp : hnSpecs ) {
+      var x = root_;
+      var pos = sp.getName().indexOf(".");
+      while ( pos > 0 ) {
```
Comment (kgrgreer) [inline src/foam/nanos/boot/Boot.java]: Easier to split().

File: src/foam/nanos/boot/NSpecFactory.java [hunk 6]
```diff
@@ -58,24 +58,27 @@ void buildService(X x) {
...
-      Object ns = ns_;
-      while ( ns != null ) {
-        if ( ns instanceof ContextAware ) {
-          ((ContextAware) ns).setX(x_.getX());
-        }
-        if ( ns instanceof NSpecAware ) {
-          if ( ((NSpecAware) ns).getNSpec() == null ) {
-            ((NSpecAware) ns).setNSpec(spec_);
+      // Check service is not a symbolic link
+      if ( ! ( ns_ instanceof X ) ) {
```
Comment (kgrgreer) [inline src/foam/nanos/boot/NSpecFactory.java]: How does this happen?

File: src/foam/nanos/boot/Boot.java (hunk 1 above)
Comment (lchanmann) [reply src/foam/nanos/boot/Boot.java:137]: Fixed in https://github.com/foam-framework/foam2/pull/5037/commits/7089465493f5f035e1ad72114e3f32d538240148.

File: src/foam/nanos/boot/Boot.java (hunk 5 above)
Comment (lchanmann) [reply src/foam/nanos/boot/Boot.java]: Fixed in https://github.com/foam-framework/foam2/pull/5037/commits/7089465493f5f035e1ad72114e3f32d538240148.

File: src/foam/nanos/boot/Boot.java (hunk 4 above)
Comment (lchanmann) [reply src/foam/nanos/boot/Boot.java]: Fixed in https://github.com/foam-framework/foam2/pull/5037/commits/7089465493f5f035e1ad72114e3f32d538240148.

File: src/foam/nanos/boot/Boot.java (hunk 3 above)
Comment (lchanmann) [reply src/foam/nanos/boot/Boot.java]: Fixed in https://github.com/foam-framework/foam2/pull/5037/commits/7089465493f5f035e1ad72114e3f32d538240148.

File: src/foam/nanos/boot/Boot.java (hunk 2 above)
Comment (lchanmann) [reply src/foam/nanos/boot/Boot.java]: Creating sub-context here polluted the root context. eg, when adding sub sub context say "bar" to "foo" sub-context which is under root, the "bar" sub-sub context also appears root.

File: src/foam/core/SubX.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package foam.core;
```
Comment (kgrgreer) [inline src/foam/core/SubX.java:7]: Copyright

Comment (lchanmann) [reply src/foam/core/SubX.java:7]: Added in https://github.com/foam-framework/foam2/pull/5037/commits/f70dab822dfbab457d3f23fe0c879fbd6619f160.

**Pull request merged**